From b99030557301dd23e23013c317b99d6d274a9895 Mon Sep 17 00:00:00 2001
From: Igor Grinberg <grinberg@compulab.co.il>
Date: Thu, 20 Dec 2012 09:49:11 +0200
Subject: ARM: i.MX6: cm-fx6: add ads7846 touchscreen

cm-fx6 has an on-board resistive touchscreen controller.
Add support fo that controller.

Signed-off-by: Igor Grinberg <grinberg@compulab.co.il>
---
 arch/arm/mach-mx6/board-cm-fx6-dl.h |    2 ++
 arch/arm/mach-mx6/board-cm-fx6-q.h  |    2 ++
 arch/arm/mach-mx6/board-cm-fx6.c    |   58 +++++++++++++++++++++++++++++++++++
 3 files changed, 62 insertions(+)

diff --git a/arch/arm/mach-mx6/board-cm-fx6-dl.h b/arch/arm/mach-mx6/board-cm-fx6-dl.h
index 3c5ce1a..1a537f8 100644
--- a/arch/arm/mach-mx6/board-cm-fx6-dl.h
+++ b/arch/arm/mach-mx6/board-cm-fx6-dl.h
@@ -25,6 +25,8 @@ static iomux_v3_cfg_t cm_fx6_dl_pads[] = {
 	MX6DL_PAD_KEY_ROW0__UART4_RXD,
 	/* Green LED */
 	MX6DL_PAD_EIM_EB3__GPIO_2_31,
+	/* TSPENDOWN */
+	MX6DL_PAD_SD4_DAT7__GPIO_2_15,
 	/* ENET */
 	MX6DL_PAD_ENET_MDIO__ENET_MDIO,
 	MX6DL_PAD_KEY_COL2__ENET_MDC,
diff --git a/arch/arm/mach-mx6/board-cm-fx6-q.h b/arch/arm/mach-mx6/board-cm-fx6-q.h
index d7a50a6..4cf0f7d 100644
--- a/arch/arm/mach-mx6/board-cm-fx6-q.h
+++ b/arch/arm/mach-mx6/board-cm-fx6-q.h
@@ -24,6 +24,8 @@ static iomux_v3_cfg_t cm_fx6_q_common_pads[] = {
 	MX6Q_PAD_KEY_ROW0__UART4_RXD,
 	/* Green LED */
 	MX6Q_PAD_EIM_EB3__GPIO_2_31,
+	/* TSPENDOWN */
+	MX6Q_PAD_SD4_DAT7__GPIO_2_15,
 	/* ENET */
 	MX6Q_PAD_ENET_MDIO__ENET_MDIO,
 	MX6Q_PAD_ENET_MDC__ENET_MDC,
diff --git a/arch/arm/mach-mx6/board-cm-fx6.c b/arch/arm/mach-mx6/board-cm-fx6.c
index 1767dec..c5446d2 100644
--- a/arch/arm/mach-mx6/board-cm-fx6.c
+++ b/arch/arm/mach-mx6/board-cm-fx6.c
@@ -31,6 +31,7 @@
 #include <linux/fsl_devices.h>
 #include <linux/spi/spi.h>
 #include <linux/spi/flash.h>
+#include <linux/spi/ads7846.h>
 #include <linux/i2c.h>
 #include <linux/i2c/pca953x.h>
 #include <linux/i2c/at24.h>
@@ -79,6 +80,7 @@
 
 /* GPIO PIN, sort by PORT/BIT */
 #define CM_FX6_GREEN_LED		IMX_GPIO_NR(2, 31)
+#define CM_FX6_TSPENDOWN		IMX_GPIO_NR(2, 15)
 #define CM_FX6_ECSPI1_CS0		IMX_GPIO_NR(2, 30)
 #define CM_FX6_ECSPI1_CS1		IMX_GPIO_NR(3, 19)
 
@@ -312,6 +314,44 @@ static struct flash_platform_data cm_fx6_spi_flash_data = {
 };
 #endif
 
+#if defined(CONFIG_TOUCHSCREEN_ADS7846) || \
+	defined(CONFIG_TOUCHSCREEN_ADS7846_MODULE)
+static struct regulator_consumer_supply ads7846_supplies[] = {
+	REGULATOR_SUPPLY("vcc", "spi0.1"),
+};
+
+static struct regulator_init_data ads7846_vcc_data = {
+	.num_consumer_supplies = ARRAY_SIZE(ads7846_supplies),
+	.consumer_supplies = ads7846_supplies,
+};
+
+static struct fixed_voltage_config ads7846_reg_config = {
+	.supply_name	= "ads7846_vcc",
+	.microvolts	= 3300000,
+	.gpio		= -EINVAL,
+	.init_data	= &ads7846_vcc_data,
+};
+
+static struct platform_device ads7846_reg_pdev = {
+	.name		= "reg-fixed-voltage",
+	.id		= 10,
+	.dev		= {
+		.platform_data = &ads7846_reg_config,
+	},
+};
+static struct ads7846_platform_data ads7846_config = {
+	.x_max			= 0x0fff,
+	.y_max			= 0x0fff,
+	.x_plate_ohms		= 180,
+	.pressure_max		= 255,
+	.debounce_max		= 30,
+	.debounce_tol		= 10,
+	.debounce_rep		= 1,
+	.gpio_pendown		= CM_FX6_TSPENDOWN,
+	.keep_vref_on		= 1,
+};
+#endif
+
 static struct spi_board_info cm_fx6_spi0_board_info[] = {
 #if defined(CONFIG_MTD_M25P80) || defined(CONFIG_MTD_M25P80_MODULE)
 	{
@@ -323,10 +363,28 @@ static struct spi_board_info cm_fx6_spi0_board_info[] = {
 		.platform_data	= &cm_fx6_spi_flash_data,
 	},
 #endif
+#if defined(CONFIG_TOUCHSCREEN_ADS7846) || \
+	defined(CONFIG_TOUCHSCREEN_ADS7846_MODULE)
+	{
+		.modalias	= "ads7846",
+		.max_speed_hz	= 1500000,
+		.bus_num	= 0,
+		.chip_select	= 1,
+		.irq		= gpio_to_irq(CM_FX6_TSPENDOWN),
+		.platform_data	= &ads7846_config,
+	},
+#endif
 };
 
 static void cm_fx6_spi_devices_init(void)
 {
+	int err;
+
+	err = platform_device_register(&ads7846_reg_pdev);
+	if (err)
+		pr_err("%s: ADS7846 regulator register failed: %d\n",
+		       __func__, err);
+
 	spi_register_board_info(cm_fx6_spi0_board_info,
 				ARRAY_SIZE(cm_fx6_spi0_board_info));
 }
-- 
1.7.9.5

